File: AIM/AIM/UI/Views/StationInfoView.m
//
//  StationInfoView.m
//  AIM
//
//  Created by Oskar Szydlowski on 05.03.2018.
//  Copyright © 2018 Oskar Szydlowski. All rights reserved.
//

#import "StationInfoView.h"

@implementation StationInfoView

-(instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:CGRectZero];
    [self setupView];
    return self;
}

-(void)setupView {
    self.backgroundColor = [UIColor whiteColor];
    [self setupSeparator];
    [self setupImageView];
    [self setupNameLabel];
    [self setupDescriptionLabel];
}

-(void)setupSeparator {
    self.separator = [UIView new];
    self.separator.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:self.separator];
    [self addConstraints:@[
                           [self.separator.leadingAnchor constraintEqualToAnchor:self.leadingAnchor],
                           [self.separator.trailingAnchor constraintEqualToAnchor:self.trailingAnchor],
                           [self.separator.bottomAnchor constraintEqualToAnchor:self.bottomAnchor],
                           [self.separator.heightAnchor constraintEqualToConstant:5.0]
    ]];
    self.separator.backgroundColor = [UIColor grayColor];
}

-(void)setupImageView {
    self.imageView = [UIImageView new];
    self.imageView.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:self.imageView];
    [self addConstraints:@[
                           [self.imageView.topAnchor constraintEqualToAnchor:self.topAnchor constant:5.0],
                           [self.imageView.leadingAnchor constraintEqualToAnchor:self.leadingAnchor constant:5.0],
                           [self.imageView.heightAnchor constraintEqualToConstant:80.0],
                           [self.imageView.widthAnchor constraintEqualToConstant:80.0]
    ]];
    self.imageView.contentMode = UIViewContentModeScaleAspectFit;
}

-(void)setupNameLabel {
    self.nameLabel = [UILabel new];
    self.nameLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:self.nameLabel];
    [self addConstraints:@[
                           [self.nameLabel.topAnchor constraintEqualToAnchor:self.topAnchor constant:5.0],
                           [self.nameLabel.leadingAnchor constraintEqualToAnchor:self.imageView.trailingAnchor constant: 10.0]
    ]];
}

-(void)setupDescriptionLabel {
    self.descriptionLabel = [UILabel new];
    self.descriptionLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self addSubview:self.descriptionLabel];
    [self addConstraints:@[
                           [self.descriptionLabel.topAnchor constraintEqualToAnchor:self.nameLabel.bottomAnchor constant:10.0],
                           [self.descriptionLabel.leadingAnchor constraintEqualToAnchor:self.nameLabel.leadingAnchor],
                           [self.descriptionLabel.trailingAnchor constraintEqualToAnchor:self.trailingAnchor constant:-5.0],
                           [self.descriptionLabel.bottomAnchor constraintEqualToAnchor:self.separator.topAnchor],
    ]];
    self.descriptionLabel.numberOfLines = 0;
}

- (void) displayImage: (NSURL *) URL {
    NSData * imageData = [[NSData alloc] initWithContentsOfURL: URL];
    [self.imageView setImage: [UIImage imageWithData:imageData]];
}

- (void) displayName: (NSString *) name {
    self.nameLabel.text = name;
}

- (void) displayDescription: (NSString *) description {
    self.descriptionLabel.text = description;
}

@end

File: AIM/AIM/UI/Views/StationInfoView.swift
//
//  StationInfoView.swift
//  AIM
//
//  Created by Oskar Szydlowski on 28.02.2018.
//  Copyright © 2018 Oskar Szydlowski. All rights reserved.
//

import UIKit

class StationInfoView: UIView {
    
    private let nameLabel = UILabel()
    private let descriptionLabel = UILabel()
    private let imageView = UIImageView()
    private let separator: UIView = {
        let separator = UIView()
        separator.backgroundColor = UIColor.gray
        return separator
    }()
    
    override init(frame: CGRect) {
        super.init(frame: .zero)
        setupView()
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func setupView() {
        backgroundColor = UIColor.white
        setupSeparator()
        setupImageView()
        setupNameLabel()
        setupDescriptionLabel()
    }
    
    private func setupSeparator() {
        addSubview(separator, with: [
            separator.leadingAnchor.constraint(equalTo: leadingAnchor),
            separator.trailingAnchor.constraint(equalTo: trailingAnchor),
            separator.bottomAnchor.constraint(equalTo: bottomAnchor),
            separator.heightAnchor.constraint(equalToConstant: 5.0)
        ])
    }
    
    private func setupImageView() {
        addSubview(imageView, with: [
            imageView.topAnchor.constraint(equalTo: topAnchor, constant: 5.0),
            imageView.leadingAnchor.constraint(equalTo: leadingAnchor, constant: 5.0),
            imageView.heightAnchor.constraint(equalToConstant: 80.0),
            imageView.widthAnchor.constraint(equalToConstant: 80.0)
        ])
    }
    
    private func setupNameLabel() {
        addSubview(nameLabel, with: [
            nameLabel.topAnchor.constraint(equalTo: topAnchor, constant: 5.0),
            nameLabel.leadingAnchor.constraint(equalTo: imageView.trailingAnchor, constant: 10.0)
        ])
    }
    
    private func setupDescriptionLabel() {
        addSubview(descriptionLabel, with: [
            descriptionLabel.topAnchor.constraint(equalTo: nameLabel.bottomAnchor, constant: 10.0),
            descriptionLabel.leadingAnchor.constraint(equalTo: nameLabel.leadingAnchor),
            descriptionLabel.trailingAnchor.constraint(equalTo: trailingAnchor, constant: -5.0),
            descriptionLabel.bottomAnchor.constraint(equalTo: separator.topAnchor)
        ])
        
        descriptionLabel.numberOfLines = 0
    }
    
    func displayImage(_ URL: URL) {
        imageView.setImage(with: URL)
    }
    
    func displayName(_ name: String) {
        nameLabel.text = name
    }
    
    func displayDescription(_ description: String) {
        descriptionLabel.text = description
    }
}
